src/Document.ts

```diff
@@ -199,9 +200,16 @@ class Document {
 						setIn(this, keyPath, value);
 
 						const errors = await schemaType.validate(value, this);
-						if (errors.length > 0) {
+						if (this.#isArrayOfStrings(errors) && errors.length > 0) {
 							documentErrors.set(keyPath, errors);
 						}
+						if (errors instanceof Map || this.#isArrayOfMaps(errors)) {
+							ensureArray(errors).forEach((errorMap) => {
+								errorMap.forEach((error, key) => {
+									documentErrors.set(`${keyPath}.${key}`, error);
+								});
+							});
+						}
 					} catch (err) {
 						// an error was thrown - return the message from that error in the documentErrors list
 						documentErrors.set(keyPath, err.message);
```

Comment:
I think the `documentErrors` type should be narrowed to `Map<string, string[]`>. It appears to me that this is the only scenario where a `string` is the `Map` value. If this set the value as `[err.message`] instead I think this type can be simplified a lot.

Reply:
The `DataValidationError` type can also have its type narrowed in the same way.